Fall of Babylon - Bug Fixes

arcadeboss opened the pull request:
This fixes two bugs related to displaying too many messages to players for the "seal" and "spy" messages. (Used 3 accounts to verify during testing)

Also removed the player name replacement logic, since it doesn't work with match-level events.

Pablete1234 commented:
> Also removed the player name replacement logic, since it doesn't work with match-level events.

looking at your current usage in the xml, it should work. Match-scoped message actions can use a player replacement, the only thing you need is the variable used to be player-scoped and exclusive

mitchts reviewed (changes requested):
The trigger for the seal broken messages is player scoped, but the actual filter that sets it off has no relation to a specific player. That could be what is being referenced.

You could instead use block drops to apply a kit when webs are broken within the wool room region. The kit has the set x_seal_breaker action as part of it. That solves getting the correct player name. In that same action, switch scope to match conditionally if the x-seal-broke filter returns true.

It looks like by changing the team upgrade alerts and spy reports to team scope you wouldn't need the cooldowns (which is a bandaid rather than addressing the issue). Currently when `<trigger filter="all(armor_upgrade_variable=1)" scope="player">` is triggered, then, for example, every single player on Blue triggers 1 spy report for Red (20 Blue -> 20 spy reports).

Pablete1234 commented:
Yeah i see the issue, those seal-broke triggers are player-scoped but they act on a match-scoped filter, meaning each player is going to trigger it for everyone. This is not reasonable behavior!!

arcadeboss commented:
> The trigger for the seal broken messages is player scoped, but the actual filter that sets it off has no relation to a specific player. That could be what is being referenced.
> 
> You could instead use block drops to apply a kit when webs are broken within the wool room region. The kit has the set x_seal_breaker action as part of it. That solves getting the correct player name. In that same action, switch scope to match conditionally if the x-seal-broke filter returns true.
> 
> It looks like by changing the team upgrade alerts and spy reports to team scope you wouldn't need the cooldowns (which is a bandaid rather than addressing the issue). Currently when `<trigger filter="all(armor_upgrade_variable=1)" scope="player">` is triggered, then, for example, every single player on Blue triggers 1 spy report for Red (20 Blue -> 20 spy reports).

I had tried using the block drops to apply a kit to solve this same issue for Bed Rush, but I wasn't able to get the kit to work.  This case is a bit different, and might work better though.  

When you say "kit has the set x_seal_breaker action as part of it", how is that implemented?  My thought previously was to give some exclusive item in the kit and then have another action that is watching for carrying that item.   Is this roughly what you mean, or is there a more direct way to do it?

Pablete1234 commented:
> Is this roughly what you mean, or is there a more direct way to do it?
there's an action kit
```xml
<kit>
  <action>
    <set   ...>
  </action>
</kit>
```xml

arcadeboss commented:
> > Is this roughly what you mean, or is there a more direct way to do it?
> 
> there's an action kit
> 
> ```
> <kit>
>   <action>
>     <set   ... />
>   </action>
> </kit>
> ```

Thank you, TIL.

arcadeboss commented:
> You could instead use block drops to apply a kit when webs are broken within the wool room region. The kit has the set x_seal_breaker action as part of it. That solves getting the correct player name. In that same action, switch scope to match conditionally if the x-seal-broke filter returns true.
> 
> It looks like by changing the team upgrade alerts and spy reports to team scope you wouldn't need the cooldowns (which is a bandaid rather than addressing the issue). Currently when `<trigger filter="all(armor_upgrade_variable=1)" scope="player">` is triggered, then, for example, every single player on Blue triggers 1 spy report for Red (20 Blue -> 20 spy reports).

I have removed all the cool-down logic, and added the block-drop -> kit -> action logic to capture the seal breaking player, this allowed me to add the player name replacement logic back in.   Also, the scope of filters reworked to avoid unneeded invocations.

mitchts commented on ctw/fall_of_babylon/map.xml:
```diff
@@ -118,6 +118,26 @@
     <kit id="gold-block" force="true" filter="always">
         <item material="gold block"/>
     </kit>
+    <kit id="lime-seal-breaker-kit" force="true">
+        <action>
+            <set var="lime_seal_breaker" value="1"/>
+        </action>
+    </kit>
+    <kit id="cyan-seal-breaker-kit" force="true">
+        <action>
+            <set var="cyan_seal_breaker" value="1"/>
+        </action>
+    </kit>
+    <kit id="orange-seal-breaker-kit" force="true">
+        <action>
+            <set var="orange_seal_breaker" value="1"/>
+        </action>
+    </kit>
+    <kit id="yellow-seal-breaker-kit" force="true">
+        <action>
+            <set var="yellow_seal_breaker" value="1"/>
+        </action>
+    </kit>
```
```suggestion
    <kit id="seal-breaker-kit" force="true">
        <action filter="lime-woolroom">
            <set var="lime_seal_breaker" value="1"/>
        </action>
        <action filter="cyan-woolroom">
            <set var="cyan_seal_breaker" value="1"/>
        </action>
        <action filter="orange-woolroom">
            <set var="orange_seal_breaker" value="1"/>
        </action>
        <action filter="yellow-woolroom">
            <set var="yellow_seal_breaker" value="1"/>
        </action>
    </kit>
```

mitchts commented on ctw/fall_of_babylon/map.xml:
```diff
@@ -152,6 +172,62 @@
+            <any>
+                <material>diamond block</material>
+            </any>
+        </filter>
+        <drops>
+            <item chance="1" amount="12" material="diamond"/>
+        </drops>
+        <region>
+            <region id="cyan-woolroom"/>
+        </region>
+        <kit id="cyan-seal-breaker-kit"/>
+    </rule>
+    <rule>
+        <filter>
+            <any>
+                <material>diamond block</material>
+            </any>
+        </filter>
+        <drops>
+            <item chance="1" amount="12" material="diamond"/>
+        </drops>
+        <region>
+            <region id="yellow-woolroom"/>
+        </region>
+        <kit id="yellow-seal-breaker-kit"/>
+    </rule>
+    <rule>
+        <filter>
+            <any>
+                <material>diamond block</material>
+            </any>
+        </filter>
+        <drops>
+            <item chance="1" amount="12" material="diamond"/>
+        </drops>
+        <region>
+            <region id="orange-woolroom"/>
+        </region>
+        <kit id="orange-seal-breaker-kit"/>
+    </rule>
```
```suggestion
    <rule region="woolrooms" kit="seal-breaker-kit">
        <filter>
            <any>
                <material>diamond block</material>
            </any>
        </filter>
        <drops>
            <item chance="1" amount="12" material="diamond"/>
        </drops>
    </rule>
```

mitchts commented on ctw/fall_of_babylon/map.xml:
```diff
@@ -2377,14 +2437,16 @@
-                <rectangle id="orange-woolroom" min="143,-146" max="116,-119"/>
-                <rectangle id="yellow-woolroom" min="-115,-146" max="-142,-119"/>
-            </union>
+            <negative id="not-woolrooms">
+                <union id="reds-woolrooms">
+                    <rectangle id="lime-woolroom" min="-142,146" max="-115,120"/>
+                    <rectangle id="cyan-woolroom" min="116,146" max="143,120"/>
+                </union>
+                <union id="blues-woolrooms">
+                    <rectangle id="orange-woolroom" min="143,-146" max="116,-119"/>
+                    <rectangle id="yellow-woolroom" min="-115,-146" max="-142,-119"/>
+                </union>
+            </negative>
```
```suggestion
            <negative id="not-woolrooms">
                <union id="woolrooms">
                    <union id="reds-woolrooms">
                        <rectangle id="lime-woolroom" min="-142,146" max="-115,120"/>
                        <rectangle id="cyan-woolroom" min="116,146" max="143,120"/>
                    </union>
                    <union id="blues-woolrooms">
                        <rectangle id="orange-woolroom" min="143,-146" max="116,-119"/>
                        <rectangle id="yellow-woolroom" min="-115,-146" max="-142,-119"/>
                    </union>
                </union>
            </negative>
```

arcadeboss replied:
> 

Thank you for these simplifications.  They have been applied now.

mitchts reviewed (commented):
Changes look good to me. Just made some suggestions to simplify some of the repeated logic.
